spec/support/rails_template.rb

```diff
@@ -91,5 +91,14 @@
   rake "parallel:drop parallel:create parallel:load_schema", env: ENV['RAILS_ENV']
 end
 
+# Add both assets to test as some specs still rely on sprockets
+generate "active_admin:assets"
+generate "active_admin:webpacker"
```

Comment:
`generate "active_admin:webpacker"` shouldn't be necessary because that's done by `generate "active_admin:install --webpacker"`, right? At least I get some "identical" messages for the assets when I run the generator, meaning that the files are already there.

Regarding regenerating sprockets assets, I think we can remove the dependency on sprockets assets from our tests and kill that too.

I made the following changes to be able to do that:

```diff
diff --git a/spec/requests/stylesheets_spec.rb b/spec/requests/stylesheets_spec.rb
index 6e67cf047..f34922a5f 100644
--- a/spec/requests/stylesheets_spec.rb
+++ b/spec/requests/stylesheets_spec.rb
@@ -1,6 +1,6 @@
 require 'rails_helper'
 
-RSpec.describe "Stylesheets", type: :request do
+RSpec.describe "Stylesheets", type: :request, unless: ActiveAdmin.application.use_webpacker do
   require "sprockets"
 
   let(:css) do
diff --git a/spec/support/rails_template.rb b/spec/support/rails_template.rb
index 0bfa5953a..5b1a3b02d 100644
--- a/spec/support/rails_template.rb
+++ b/spec/support/rails_template.rb
@@ -1,7 +1,15 @@
 # Rails template to build the sample app for specs
 
-create_file 'app/assets/stylesheets/some-random-css.css'
-create_file 'app/assets/javascripts/some-random-js.js'
+webpacker_app = ENV["BUNDLE_GEMFILE"] == File.expand_path("../../gemfiles/rails_60_webpacker/Gemfile", __dir__)
+
+if webpacker_app
+  create_file 'app/javascript/packs/some-random-css.css'
+  create_file 'app/javascript/packs/some-random-js.js'
+else
+  create_file 'app/assets/stylesheets/some-random-css.css'
+  create_file 'app/assets/javascripts/some-random-js.js'
+end
+
 create_file 'app/assets/images/a/favicon.ico'
 
 require 'active_admin/dependency'
diff --git a/spec/unit/generators/install_spec.rb b/spec/unit/generators/install_spec.rb
index 0798dd31f..1362fa499 100644
--- a/spec/unit/generators/install_spec.rb
+++ b/spec/unit/generators/install_spec.rb
@@ -3,13 +3,21 @@ require 'rails_helper'
 RSpec.describe "AA installation" do
   context "should create" do
     it "active_admin.scss" do
-      path = Rails.root + "app/assets/stylesheets/active_admin.scss"
+      path = if ActiveAdmin.application.use_webpacker
+               Rails.root + "app/javascript/stylesheets/active_admin.scss"
+             else
+               Rails.root + "app/assets/stylesheets/active_admin.scss"
+             end
 
       expect(File.exist?(path)).to eq true
     end
 
     it "active_admin.js" do
-      path = Rails.root + "app/assets/javascripts/active_admin.js"
+      path = if ActiveAdmin.application.use_webpacker
+               Rails.root + "app/javascript/packs/active_admin.js"
+             else
+               Rails.root + "app/assets/javascripts/active_admin.js"
+             end
 
       expect(File.exist?(path)).to eq true
     end
```

With those changes, everything becomes green for me, except for 

```

Failures:

  1) ActiveAdmin::Views::Pages::Layout should be the default translation
     Failure/Error: stylesheet_tag = active_admin_namespace.use_webpacker ? stylesheet_pack_tag(style, options) : stylesheet_link_tag(style, options)
     
     Sprockets::Rails::Helper::AssetNotFound:
       The asset "active_admin.css" is not present in the asset pipeline.
     # ./lib/active_admin/views/pages/base.rb:32:in `block (2 levels) in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:31:in `each'
     # ./lib/active_admin/views/pages/base.rb:31:in `block in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:26:in `build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:8:in `build'
     # ./spec/unit/views/pages/layout_spec.rb:30:in `block (3 levels) in <top (required)>'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `new'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `arbre'
     # ./spec/support/active_admin_integration_spec_helper.rb:40:in `render_arbre_component'
     # ./spec/unit/views/pages/layout_spec.rb:29:in `block (2 levels) in <top (required)>'
     # ./spec/unit/views/pages/layout_spec.rb:43:in `block (2 levels) in <top (required)>'

  2) ActiveAdmin::Views::Pages::Layout should have lang attribute on the html element
     Failure/Error: stylesheet_tag = active_admin_namespace.use_webpacker ? stylesheet_pack_tag(style, options) : stylesheet_link_tag(style, options)
     
     Sprockets::Rails::Helper::AssetNotFound:
       The asset "active_admin.css" is not present in the asset pipeline.
     # ./lib/active_admin/views/pages/base.rb:32:in `block (2 levels) in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:31:in `each'
     # ./lib/active_admin/views/pages/base.rb:31:in `block in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:26:in `build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:8:in `build'
     # ./spec/unit/views/pages/layout_spec.rb:30:in `block (3 levels) in <top (required)>'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `new'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `arbre'
     # ./spec/support/active_admin_integration_spec_helper.rb:40:in `render_arbre_component'
     # ./spec/unit/views/pages/layout_spec.rb:29:in `block (2 levels) in <top (required)>'
     # ./spec/unit/views/pages/layout_spec.rb:47:in `block (2 levels) in <top (required)>'

  3) ActiveAdmin::Views::Pages::Layout should be the @page_title if assigned in the controller
     Failure/Error: stylesheet_tag = active_admin_namespace.use_webpacker ? stylesheet_pack_tag(style, options) : stylesheet_link_tag(style, options)
     
     Sprockets::Rails::Helper::AssetNotFound:
       The asset "active_admin.css" is not present in the asset pipeline.
     # ./lib/active_admin/views/pages/base.rb:32:in `block (2 levels) in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:31:in `each'
     # ./lib/active_admin/views/pages/base.rb:31:in `block in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:26:in `build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:8:in `build'
     # ./spec/unit/views/pages/layout_spec.rb:30:in `block (3 levels) in <top (required)>'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `new'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `arbre'
     # ./spec/support/active_admin_integration_spec_helper.rb:40:in `render_arbre_component'
     # ./spec/unit/views/pages/layout_spec.rb:29:in `block (2 levels) in <top (required)>'
     # ./spec/unit/views/pages/layout_spec.rb:37:in `block (2 levels) in <top (required)>'

  4) ActiveAdmin::Views::Pages::Layout the body should have namespace class
     Failure/Error: stylesheet_tag = active_admin_namespace.use_webpacker ? stylesheet_pack_tag(style, options) : stylesheet_link_tag(style, options)
     
     Sprockets::Rails::Helper::AssetNotFound:
       The asset "active_admin.css" is not present in the asset pipeline.
     # ./lib/active_admin/views/pages/base.rb:32:in `block (2 levels) in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:31:in `each'
     # ./lib/active_admin/views/pages/base.rb:31:in `block in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:26:in `build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:8:in `build'
     # ./spec/unit/views/pages/layout_spec.rb:30:in `block (3 levels) in <top (required)>'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `new'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `arbre'
     # ./spec/support/active_admin_integration_spec_helper.rb:40:in `render_arbre_component'
     # ./spec/unit/views/pages/layout_spec.rb:29:in `block (2 levels) in <top (required)>'
     # ./spec/unit/views/pages/layout_spec.rb:56:in `block (3 levels) in <top (required)>'

  5) ActiveAdmin::Views::Pages::Layout the body should have class 'active_admin'
     Failure/Error: stylesheet_tag = active_admin_namespace.use_webpacker ? stylesheet_pack_tag(style, options) : stylesheet_link_tag(style, options)
     
     Sprockets::Rails::Helper::AssetNotFound:
       The asset "active_admin.css" is not present in the asset pipeline.
     # ./lib/active_admin/views/pages/base.rb:32:in `block (2 levels) in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:31:in `each'
     # ./lib/active_admin/views/pages/base.rb:31:in `block in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:26:in `build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:8:in `build'
     # ./spec/unit/views/pages/layout_spec.rb:30:in `block (3 levels) in <top (required)>'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `new'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `arbre'
     # ./spec/support/active_admin_integration_spec_helper.rb:40:in `render_arbre_component'
     # ./spec/unit/views/pages/layout_spec.rb:29:in `block (2 levels) in <top (required)>'
     # ./spec/unit/views/pages/layout_spec.rb:52:in `block (3 levels) in <top (required)>'

Finished in 12.15 seconds (files took 8.1 seconds to load)
120 examples, 5 failures

Failed examples:

rspec ./spec/unit/views/pages/layout_spec.rb:40 # ActiveAdmin::Views::Pages::Layout should be the default translation
rspec ./spec/unit/views/pages/layout_spec.rb:46 # ActiveAdmin::Views::Pages::Layout should have lang attribute on the html element
rspec ./spec/unit/views/pages/layout_spec.rb:34 # ActiveAdmin::Views::Pages::Layout should be the @page_title if assigned in the controller
rspec ./spec/unit/views/pages/layout_spec.rb:55 # ActiveAdmin::Views::Pages::Layout the body should have namespace class
rspec ./spec/unit/views/pages/layout_spec.rb:51 # ActiveAdmin::Views::Pages::Layout the body should have class 'active_admin'
```

Those failures I managed to fix them by moving the `use_webpacker` setting from `namespace_settings.rb` to `application_settings.rb` and using `active_admin_application.use_webpacker` instead of `active_admin_namespace.user_webpacker` when rendering the packs inside `lib/active_admin/views/pages/base.rb`.

But that changes how this PR works because right now `use_webpacker` is toggleable at the namespace level, and with those changes it would only be toggleable for the whole activeadmin application.

However, I think it should be possible to change the test appropriately to get it to pass with the current implementation.

Reply:
Indeed, it seems that I couldn't get that to work - as we were having sprockets related tests with webpacker app. Checking `use_webpacker` is a good option to me 👍 
With those changes, no need to generate both confs indeed 👌

Reply:
I slightly changed your suggestion to not test `register_stylesheet` things at all while testing webpacker. I actually doubt it would make any sense to do so anyway as a webpacker user would play with packs instead of registering extra stylesheets. 
I'd more likely deprecate `register_stylesheet` when using webpacker, what do you think?

Reply:
mmm, not sure but I think that grants a separate discussion for which I need to learn more about all this :sweat_smile:. But for now we're using `register_stylesheet` to register the `active_admin/prints.scss` stylesheet even when using webpacker, so I think it makes sense to test it?

Reply:
Forget it, you're right :)

Reply:
Ok I've managed to make it work while keeping the setting at namespace level. Thanks for the input here 👍